pkg/composition/controller2.go

```diff
@@ -0,0 +1,624 @@
+package composition
```

Comment:
Here's the rewrite. Assume it's new code your reviewing from scratch. Refer to the old code if you're curious about something but otherwise focus on the structure of this code.